src/core/iop/memcard.cpp

```diff
@@ -1,6 +1,308 @@
+			this->formatted = false;
+			break;
+		}
+	}
+
+	// Check if backup block 2 (labelled 2 in documentation, but actually the
+	// first on the card physically) is not erased. Page 16352 is the first page
+	// of backup block 2.
+	if (!is_erase_block_erased(16352 * bytes_per_page))
+	{
+		printf("[MCD] (WARN) Backup block 2 was NOT erased when this memcard was opened! The last write op on this memcard may have failed! The emulated PS2 should be restoring the old data soon, if it hasn't already...");
+	}
+
+	return true;
+}
+
+// Read length bytes from memcard starting at address into destination.
+bool Memcard::read(std::vector<uint8_t> destination, uint32_t address, uint32_t length)
+{
+	if (!address_in_bounds(address, length))
+	{
+		printf("[MCD] (WARN) Attempted to read out-of-bounds memcard address %d! Aborting...", address);
+		return false;
+	}
+
+	memcpy(destination.data(), &raw_data[address], length);
+	return true;
+}
+
+// Write length bytes from source starting at address into memcard.
+bool Memcard::write(std::vector<uint8_t> source, uint32_t address)
+{
+	if (!address_in_bounds(address, source.size()))
+	{
+		printf("[MCD] (WARN) Attempted to write out-of-bounds memcard address %d! Aborting...", address);
+		return false;
+	}
+
+	// Check if a bitwise and of source and address bytes would give a result
+	// different from the source byte. If so, that means something else
+	// currently exists here, or the erase block containing address has not been
+	// wiped yet. 
+	for (int i = 0; i < source.size(); i++)
+	{
+		if (raw_data[address + i] & source[i] != source[i])
+		{
+			printf("[MCD] (WARN) Attempted to write to memcard address %d, but its erase block hasn't been erased yet! Aborting...", address + i);
+			return false;
+		}
+	}
+
+	// Info printf's for when an important sector is written
+	if (address < 528) // Superblock
+	{
+		printf("[MCD] (INFO) Memcard write op on superblock (addr = %d, size = %d)", address, source.size());
+	}
+	else if (address < checksum_address + 8) // Checksum
+	{
+		printf("[MCD] (INFO) Memcard write op on checksum (addr = %d, size = %d)", address, source.size());
+	}
+	else if (address < 8448) // Unused space
+	{
+		printf("[MCD] (INFO) Memcard write op on unused card space (addr = %d, size = %d)", address, source.size());
+	}
+	else if (address < 9504) // Indirect FAT
+	{
+		printf("MCD] (INFO) Memcard write op on indirect FAT (addr = %d, size = %d)", address, source.size());
+	}
+	else if (address < 43296) // FAT
+	{
+		printf("MCD] (INFO) Memcard write op on FAT (addr = %d, size = %d)", address, source.size());
+	}
+
+	std::ofstream file(this->file_name);
+
+	if (!file.is_open())
+	{
+		printf("[MCD] (WARN) Attempted to open ofstream on %s, but it cannot be opened! Aborting...", this->file_name);
+		return false;
+	}
+
+	memcpy(&raw_data[address], source.data(), source.size());
+	file.seekp(address);
+	file.write((char*) source.data(), source.size());
```

Comment:
Consider replacing the C-style cast with an explicit cpp cast, e.g. `static_cast<char*>(source.data())`